src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -0,0 +1,212 @@
+    std::unique_lock<std::mutex> guard(rpc_clients_mutex_);
+    actor_states_.erase(actor_id);
+    actor_states_.emplace(
+        actor_id,
+        ActorStateData(actor_data.state(), actor_data.ip_address(), actor_data.port()));
+
+    if (actor_data.state() == ActorTableData::ALIVE) {
+      // Check if this actor is the one that we're interested, if we already have
+      // a connection to the actor, or have pending requests for it, we should
+      // create a new connection.
+      if (rpc_clients_.count(actor_id) > 0 || pending_requests_.count(actor_id) > 0) {
+        HandleActorAlive(actor_id, actor_data.ip_address(), actor_data.port());
+      }
+    }
+
+    RAY_LOG(INFO) << "received notification on actor, state="
+                  << static_cast<int>(actor_data.state()) << ", actor_id: " << actor_id
+                  << ", ip address: " << actor_data.ip_address()
+                  << ", port: " << actor_data.port();
+  };
+
+  return gcs_client_.Actors().AsyncSubscribe(actor_notification_callback, nullptr);
+}
+
+void CoreWorkerDirectActorTaskSubmitter::HandleActorAlive(const ActorID &actor_id,
+                                                          std::string ip_address,
+                                                          int port) {
+  std::unique_ptr<rpc::DirectActorClient> grpc_client(
+      new rpc::DirectActorClient(ip_address, port, client_call_manager_));
+  // replace old rpc client if it exists.
+  rpc_clients_[actor_id] = std::move(grpc_client);
+
+  // Submit all pending requests.
+  auto &client = rpc_clients_[actor_id];
+  auto &requests = pending_requests_[actor_id];
+  while (!requests.empty()) {
+    const auto &request = *requests.front();
+    auto status =
+        PushTask(*client, *request.request_, request.task_id_, request.num_returns_);
+    requests.pop_front();
+  }
+}
+
+Status CoreWorkerDirectActorTaskSubmitter::PushTask(rpc::DirectActorClient &client,
+                                                    const rpc::PushTaskRequest &request,
+                                                    const TaskID &task_id,
+                                                    int num_returns) {
+  auto status = client.PushTask(
+      request,
+      [this, task_id, num_returns](Status status, const rpc::PushTaskReply &reply) {
+        if (!status.ok()) {
+          TreatTaskAsFailed(task_id, num_returns, rpc::ErrorType::ACTOR_DIED);
+          return;
+        }
+        for (int i = 0; i < reply.return_objects_size(); i++) {
+          const auto &return_object = reply.return_objects(i);
+          ObjectID object_id = ObjectID::FromBinary(return_object.object_id());
+          std::shared_ptr<LocalMemoryBuffer> data_buffer;
+          if (return_object.data().size() > 0) {
+            data_buffer = std::make_shared<LocalMemoryBuffer>(
+                const_cast<uint8_t *>(
+                    reinterpret_cast<const uint8_t *>(return_object.data().data())),
```

Comment:
Will it be better to change the type of `data` from `uint8_t *` to `void *` or `char *`? then we don't need to do this cast everywhere. We can do that in next PR if needed.

Reply:
right... let's do this in the next pr?

Reply:
sure, I was just wondering the pros and cons of those 2 ways.